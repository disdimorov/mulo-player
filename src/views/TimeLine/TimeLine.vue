<template>
  <div class="time-line">
    <div
      ref="progress"
      class="time-line__progress"
      title="Time played : Total time"
      @click="locate">
      <div
        :style="lineStyle"
        class="time-line__line"/>
    </div>
    <div class="time-line__timers">
      <div class="time-line__current">{{ this.currentSeconds | convertTime }}</div>
      <div class="time-line__total">{{ this.durationSeconds | convertTime }}</div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'TimeLine',

  props: {
    currentSeconds: {
      type: Number,
      required: true,
    },

    durationSeconds: {
      type: Number,
      required: true,
    },

    playing: {
      type: Boolean,
      required: true,
    },
  },

  computed: {
    lineStyle() {
      return { width: `${this.currentSeconds / this.durationSeconds * 100}%` };
    },
  },

  filters: {
    convertTime(val) {
      const time = new Date(val * 1000).toJSON().substr(11, 8);

      return time.indexOf('00:') === 0 ? time.substr(3) : time;
    },
  },

  methods: {
    locate(event) {
      if (!this.playing) {
        return;
      }
      const { left, width } = this.$refs.progress.getBoundingClientRect();
      const position = (event.clientX - left) / width;

      this.$emit('rewind', position);
    },
  },
};
</script>

<style scoped lang="less">
@import '~@/style/variables.less';

.time-line {
  border-right: 1px solid @player-border-color;

  &__progress {
    position: relative;
    z-index: 0;
    width: 200px;
    height: 50%;
    background-color: @player-progress-color;
    cursor: pointer;
  }
  &__line {
    position: absolute;
    z-index: 1;
    top: 0;
    left: 0;
    bottom: 0;
    background-color: @player-line-color;
  }

  &__timers {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: flex-start;
    flex-wrap: wrap;
  }

  &__current {
    padding-left: 5px;
    font-weight: 700;
  }

  &__total {
    padding-right: 5px;
    opacity: 0.5;
  }
}
</style>
