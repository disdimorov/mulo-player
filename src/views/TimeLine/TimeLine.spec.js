import { mount, shallowMount, createLocalVue } from '@vue/test-utils';
import TimeLine from './TimeLine.vue';

const localVue = createLocalVue();

describe('Component TimeLine', () => {
  const defaultProps = {
    currentSeconds: 112,
    durationSeconds: 4932,
    playing: false,
  };
  const createWrapper = ({ isWithoutStubs = false, propsData = {}, ...options } = {}) => {
    const renderer = isWithoutStubs ? mount : shallowMount;
    const wrapper = renderer(TimeLine, {
      localVue,
      propsData: Object.assign({}, defaultProps, propsData),
      ...options,
    });

    return { wrapper };
  };

  describe('snapshots', () => {
    it('should match snapshot', () => {
      const { wrapper } = createWrapper({ isWithoutStubs: true });

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('event', () => {
    it('should on click over progress bar emit rewind event with position', () => {
      const { wrapper } = createWrapper();

      wrapper.vm.$refs.progress.getBoundingClientRect = () => ({ left: 0, width: 200 });
      wrapper.find('.time-line__progress').trigger('click', { clientX: 20 });
      expect(wrapper.emitted().rewind).toEqual([[0.1]]);
    });
  });
});
