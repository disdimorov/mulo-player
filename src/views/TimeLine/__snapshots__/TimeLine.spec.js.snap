// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component TimeLine snapshots should match snapshot 1`] = `
<div class="time-line">
  <div title="Time played : Total time" class="time-line__progress">
    <div class="time-line__line" style="width: 2.2708840227088403%;"></div>
  </div>
  <div class="time-line__timers">
    <div class="time-line__current">01:52</div>
    <div class="time-line__total">01:22:12</div>
  </div>
</div>
`;
